### PR TITLE
feat: refactoring and unitTests

### PR DESCRIPTION
**Pull request by NektariosFifes**
**Description**

- Refactored ScenarioParser
- Added related UnitTests

**Related issue(s)**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,14 +1,14 @@
 {
   "name": "@asyncapi/fluffy-robot",
-  "version": "0.3.0",
+  "version": "0.2.0",
```
Comment (jonaslagoni) [inline package.json:3]: This should not be manually altered.

File: src/parser/tests/parser.test.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
+const assert = require('assert');
+const expect = require('chai').expect;
+const path = require('path');
+const expectedOutputs = require('./expectedOutputs');
+const {parseFiles} = require('../parseFiles');
+describe('Parser Tests',function() {
+  let asyncApi;
+  it('1# Should be able to parse correctly formatted asyncApi and scenario files.',async function() {
```
Comment (jonaslagoni) [inline src/parser/tests/parser.test.js:8]: You dont need the `1#` for the tests 🙂 Same for `2#`

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
Maybe it makes more sense to place the files for tests as the following?
- files
  - wrong
    - IncorrectlyFormattedScenario.yaml 
  - correct
    - ...

File: package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (NektariosFifes) [reply package.json:3]: i swear i did not change that xD.

File: package.json [hunk 3]
```diff
@@ -36,20 +36,29 @@
...
+    "js-yaml": "^3.14.1",
+    "minimatch": "^3.0.4"
   },
   "devDependencies": {
     "@semantic-release/commit-analyzer": "^8.0.1",
     "@semantic-release/github": "^7.0.4",
     "@semantic-release/npm": "^7.0.6",
     "@semantic-release/release-notes-generator": "^9.0.1",
+    "@types/assert": "^1.5.5",
     "@types/node": "^14.17.5",
```
[the remark this answers is not in the document]
Comment (NektariosFifes) [reply package.json:49]: at first i thought it was for assert module.

File: package.json [hunk 4]
```diff
@@ -36,20 +36,29 @@
...
     "all-contributors-cli": "^6.14.2",
+    "chai": "^4.3.4",
+    "chai-as-promised": "^7.1.1",
     "conventional-changelog-conventionalcommits": "^4.4.0",
     "eslint": "^7.31.0",
     "eslint-plugin-mocha": "^9.0.0",
     "eslint-plugin-security": "^1.4.0",
     "eslint-plugin-sonarjs": "^0.8.0-125",
+    "fs": "^0.0.1-security",
```
[the remark this answers is not in the document]
Comment (NektariosFifes) [reply package.json:58]: its absolutely essential 😆.

[the remark this answers is not in the document]
Comment (NektariosFifes) [reply package.json:58]: dont really know how i ended up typing npm i fs . 😆

File: src/bin/cli.js [hunk 5]
```diff
@@ -176,5 +177,12 @@ const verifyInputGetData =  async (rd, asyncApiFilepath,scenarioFile,basedir) =>
...
   await manager.createReqHandler(structuredData);
   await manager.startOperations();
-}());
+};
+
+function Main () {
+  cli();
+};
+
+Main.prototype = EventEmitter.prototype;
```
[the remark this answers is not in the document]
Comment (NektariosFifes) [reply src/bin/cli.js:186]: i wanted to make the Main function able to emit events but eventually that was not utilized

File: src/parser/tests/parser.test.js (hunk 2 above)
Comment (NektariosFifes) [reply src/parser/tests/parser.test.js:8]: i think that could be usefull tho if sm1 want to refer to a test.

File: package.json [hunk 6]
```diff
@@ -36,20 +36,29 @@
...
+    "ignore": "^4.0.6",
+    "mocha": "^9.0.3",
+    "path": "^0.12.7",
```
[the remark this answers is not in the document]
Comment (NektariosFifes) [reply package.json:61]: you mean ignore? i installed it to test some regex expressions in npmignore. i will remove it

File: package.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (jonaslagoni) [reply package.json:49]: It is both actually, dont think you need `@types` dependencies if you dont use TypeScript 🙂

File: package.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (jonaslagoni) [reply package.json:61]: Not ignore, `path`

File: src/bin/cli.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (jonaslagoni) [reply src/bin/cli.js:186]: If it is not utilized, dont add it 🙂

File: src/parser/tests/parser.test.js (hunk 2 above)
Comment (jonaslagoni) [reply src/parser/tests/parser.test.js:8]: I have never had the use-case 😄 Usually when you refer to talk about which tests are added you read it as `Parser Tests Should be able to parse ....`. It ruins the flow if `#1` is present 🙂

Comment (NektariosFifes) [reply src/parser/tests/parser.test.js:8]: ok

**Comment by NektariosFifes** [comment]
@jonaslagoni  i think this fixes the issues

**Comment by jonaslagoni** [comment]
> @jonaslagoni i think this fixes the issues

@NektariosFifes please resolve all the comments, once you have addressed them, I can see at least one is not 🙂

File: src/parser/GenerateOperations.js [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+/**
+ * Take a parsed asyncApi and scenario file and generates two objects that describe
+ * publish and subscribe operations with the needed details for the request handler
+ * be able to make requests.
+ * @param asyncApiParsed
+ * @param scenarioParsed
+ * @returns {({groupOps: {}, soloOps: {}} | {groupOps: {}, soloOps: {}})[]}
+ */
+const GenerateOperations = (asyncApiParsed,scenarioParsed) => {
+  const PublishOperations = {
```
[the remark this answers is not in the document]
Comment (NektariosFifes) [reply src/parser/GenerateOperations.js:61]: i changed that and still does not show the outdated tag>>???
some error with github
@jonaslagoni

[the remark this answers is not in the document]
Comment (NektariosFifes) [reply src/parser/GenerateOperations.js:61]: i u see in my fork it has changed that

**Comment by NektariosFifes** [comment]
@jonaslagoni  ok

File: package.json [hunk 8]
```diff
@@ -36,20 +36,24 @@
     "async-mqtt": "^2.6.1",
     "chalk": "^4.1.0",
     "commander": "^7.2.0",
-    "js-yaml": "^3.14.1"
+    "cross-env": "^7.0.3",
+    "js-yaml": "^3.14.1",
+    "minimatch": "^3.0.4"
```
Comment (jonaslagoni) [inline package.json:41]: Where do you use this dependency?

File: src/bin/cli.js [hunk 9]
```diff
@@ -176,5 +176,10 @@ const verifyInputGetData =  async (rd, asyncApiFilepath,scenarioFile,basedir) =>
...
   await manager.createReqHandler(structuredData);
   await manager.startOperations();
-}());
+};
+
+function Main () {
+  cli();
+}
 
+Main();
```
Comment (jonaslagoni) [inline src/bin/cli.js:185]: Why the wrapper function? Is it not easier just to do
```suggestion
cli();
```

File: src/parser/parseFiles.js [hunk 10]
```diff
@@ -0,0 +1,43 @@
+const parser = require('@asyncapi/parser');
+const filesystem = require('fs');
+const Ajv = require('ajv');
+const scenarioSpecs = require('../Schema');
+const yamlParser  = require('js-yaml');
+/**
+ * Parses asyncApi and scenario files and
```
Comment (jonaslagoni) [inline src/parser/parseFiles.js:7]: `and` what 😄 ?

File: src/parser/GenerateOperations.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (jonaslagoni) [reply src/parser/GenerateOperations.js:61]: Not sure what you mean here, if the branch that this PR is connected to gets commits pushed, they should be reflected here. Do you still have this problem?

Also, this has not been addressed, don't resolve any conversation that is not addressed, only those that are completely resolved.

File: src/parser/GenerateOperations.js [hunk 11]
```diff
@@ -0,0 +1,78 @@
+// eslint-disable-next-line sonarjs/cognitive-complexity
+function operationsFromChannels (channels,PublishOperations,SubscribeOperations) {
```
Comment (jonaslagoni) [inline src/parser/GenerateOperations.js:2]: Why do you change between camel-cased variables and parameters to pascal-cased? 🤔 

- https://github.com/asyncapi/fluffy-robot/pull/39/files#r687726334

**Review by jonaslagoni — CHANGES_REQUESTED** [review]

File: src/parser/GenerateOperations.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (NektariosFifes) [reply src/parser/GenerateOperations.js]: I mean that in the main thread in the PR it shows const generateOperations = (   , as const GenerateOperations = (

[the remark this answers is not in the document]
Comment (jonaslagoni) [reply src/parser/GenerateOperations.js]: I still see many cases of this behavior: https://github.com/asyncapi/simulator/blob/fd99cf3f5d16755e903f885a59529546cb7bdd62/src/parser/index.js

**Comment by NektariosFifes** [comment]
@jonaslagoni is it ok now?

**Comment by jonaslagoni** [comment]
> @jonaslagoni is it ok now?

@NektariosFifes you still have comments that are not addressed, so I assume no.

Based on the comments received, always make sure to manually walk through each change here: https://github.com/asyncapi/simulator/pull/39/files

To see if something else needs to change as well.

**Comment by NektariosFifes** [comment]
@jonaslagoni  ok i fixed the camelCase and added related eslint rule. i think its ok now XD sry for being annoying

File: src/parser/tests/operationsGenerator.test.js [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+        }
+      }
+    },
+    groupOps: {}
+  };
+  const expectedSubscribeOperations = {
+    soloOps: {},
+    groupOps: {}
+  };
+  it('Should correctly generate Publish Operations.', function () {
```
Comment (jonaslagoni) [inline src/parser/tests/operationsGenerator.test.js:65]:
```suggestion
  it('Should correctly generate publish operations', function () {
```

File: src/parser/tests/operationsGenerator.test.js [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+    [publishOperations,subscribeOperations] = generateOperations(parsedAsyncApi,parsedScenario);
+    expect(publishOperations).to.deep.equal(expectedPublishOperations);
+  });
+  it('Should correctly generate subscribe Operations.', function () {
```
Comment (jonaslagoni) [inline src/parser/tests/operationsGenerator.test.js:69]:
```suggestion
  it('Should correctly generate subscribe operations', function () {
```

File: src/parser/tests/parser.test.js [hunk 14]
```diff
@@ -0,0 +1,27 @@
+const assert = require('assert');
+const expect = require('chai').expect;
+const path = require('path');
+const expectedOutputs = require('./expectedOutputs');
+const {parseFiles} = require('../parseFiles');
+describe('Parser Tests',function() {
+  let asyncApi;
+  it('Should be able to parse correctly formatted asyncApi and scenario files.',async function() {
```
Comment (jonaslagoni) [inline src/parser/tests/parser.test.js:8]:
```suggestion
  it('Should be able to parse correctly formatted AsyncAPI and scenario files', async function() {
```

File: src/parser/tests/parser.test.js [hunk 15]
```diff
@@ -0,0 +1,27 @@
+const assert = require('assert');
+const expect = require('chai').expect;
+const path = require('path');
+const expectedOutputs = require('./expectedOutputs');
+const {parseFiles} = require('../parseFiles');
+describe('Parser Tests',function() {
```
Comment (jonaslagoni) [inline src/parser/tests/parser.test.js:6]:
```suggestion
describe('Parser', function() {
```

File: src/parser/tests/parser.test.js [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+  let asyncApi;
+  it('Should be able to parse correctly formatted asyncApi and scenario files.',async function() {
+    try {
+      [asyncApi] = await parseFiles(path.resolve(__dirname, './files/correct/CorrectlyFormattedAsyncApi.yaml'), path.resolve(__dirname, './files/correct/CorrectlyFormattedScenario.yaml'));
+    } catch (err) {
+      assert.fail('Test failed. Parser was not able to parse one of the files you provided');
+    }
+    expect(asyncApi._json).to.deep.include(expectedOutputs.p1AsyncApi._json);
+  });
+  it('Throws Error when parsing incorrectly formatted AsyncApi file.',async function () {
```
Comment (jonaslagoni) [inline src/parser/tests/parser.test.js:16]:
```suggestion
 it('should throw error when parsing incorrectly formatted AsyncAPI file', async function () {
```

File: src/parser/tests/parser.test.js [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+    let Error = {};
```
Comment (jonaslagoni) [inline src/parser/tests/parser.test.js:17]: Another upper-case

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
> @jonaslagoni ok i fixed the camelCase and added related eslint rule.

👍

You are not, just keep in mind to check all changes yourself for repeating suggestions, so you make sure a review is fully addressed before requesting another 🙂 

Got a few last couple of spelling nitpicks 😄

**Review by jonaslagoni — APPROVED** [review]
Looks good to me, good job 👍

**Pull request merged**